web/hooks/use-subscription.ts

```diff
@@ -0,0 +1,253 @@
+import {
+  SupabaseClient,
+  TableName,
+  RowFor,
+  DataFor,
+  primaryKeyColumnsByTable,
+} from 'common/supabase/utils'
+import { useEffect, useId, useRef, useState } from 'react'
+import { uniqBy } from 'lodash'
+import { filterDefined } from 'common/util/array'
+import { RealtimeChannel } from '@supabase/realtime-js'
+
+//TODO: fs_updated_time relies on firebase, which we are moving off of. We
+// have no better solution for updated time atm.
+function hasFsUpdatedTime(obj: any): obj is { fs_updated_time: number } {
+  return 'fs_updated_time' in obj
+}
+type RowFilter<T extends TableName> = { k: keyof RowFor<T>; v: string }
+const CHANNEL_NAME = 'supabase-realtime-values'
+
+let mainChannel: undefined | RealtimeChannel = undefined
```

Comment:
Doesn't this imply there's only one channel? What if you have two hooks listening to different tables?

Reply:
The channel has to restart and subscribe to updates to the two different tables. We should minimize channels per client due to realtime limitations:
![Screenshot 2023-04-26 at 2 56 16 PM](https://user-images.githubusercontent.com/23196210/234700424-9f425323-218b-44cd-bf3a-6117e394c254.png)